### PR TITLE
[docs-only] Docs storage backend cephfs

### PR DESCRIPTION
**Pull request by butonic**
@dragotin @micbar @C0rby @refs @aduffeck Just a heads up on what I found upstream and where I see room for imprevement.

### CODE REVIEW

File: docs/ocis/storage-backends/cephfs.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+## Future work
+- The spaces concept matches cephfs subvolumes. We can implement the CreateStorageSpace call with that, keep track of the list of storage spaces using symlinks, like for the id based lookup.
+- The share manager needs a persistence layer.
+- Currently we persist using a single json file.
+- As it basically provides two lists, *shared with me* and *shared with others*, we could persist them directly on cephfs!
+  - If needed for redundancy, the share manager can be run multiple times, backed by the same cephfs
+  - To save disk io the data can be cached in memory, and invalidated using stat requests.
+- A good tradeoff would be a folder for each user with a json file for each list. That way, we only have to open and read a single file when the user want's to list the shares.    
+- To allow deprovisioning a user the data should by sharded by userid.
```
Comment (refs) [inline docs/ocis/storage-backends/cephfs.md:55]: Could you elaborate here? How is deprovisioning a user directly linked with sharding?

**Review by refs — CHANGES_REQUESTED** [review]
Just a few minor typos, and a question. Other than that thanks for documenting the CephFS future work!

**Review by micbar — APPROVED** [review]

**Review by refs — APPROVED** [review]

File: docs/ocis/storage-backends/cephfs.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+## Development
+
+The cephfs development happens in a [reva branch](https://github.com/cs3org/reva/pull/1209) and is currently driven by CERN. 
+
+## Architecture
+
+In the original approach the driver was based on the localfs driver, relying on a locally mounted cephfs. It would interface with it using the POSIX apis. This has been changed to directly call the Ceph API using https://github.com/ceph/go-ceph. It allows using the ceph admin APIs to create subvolumes for user homes and maintain a file id to path mapping using symlinks.
+
+## Implemented Aspects
+The recursive change time built ino cephfs is used to implement the etag propagation expected by the ownCloud clients. This allows oCIS to pick up changes that have been made by external tools, bypassing any oCIS APIs.
```
Comment (dragotin) [inline docs/ocis/storage-backends/cephfs.md:23]: Hm, this is a bit brief - it requires knowledge about Ceph etc. Is there any high level documentation from the Ceph side available? Maybe add a link?

File: docs/ocis/storage-backends/cephfs.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+.
+├── reva
+│   └── einstein
+│       ├── Pictures
+│       └── welcome.txt
+└── .reva_hidden
+    ├── .fileids
+    │   ├── 50BC39D364A4703A20C58ED50E4EADC3_570078 -> /tmp/cephfs/reva/einstein
+    │   ├── 571EFB3F0ACAE6762716889478E40156_570081 -> /tmp/cephfs/reva/einstein/Pictures
+    │   └── C7A1397524D0419B38D04D539EA531F8_588108 -> /tmp/cephfs/reva/einstein/welcome.txt
```
Comment (dragotin) [inline docs/ocis/storage-backends/cephfs.md:37]: Please do not use `/tmp` here - that confuses people who know the linux filesystem layout (I assume these files are *not* temporary). The right location would be `/var/lib/reva/` or another "neutral" dir like `/data`.

File: docs/ocis/storage-backends/cephfs.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+The cephfs development happens in a [reva branch](https://github.com/cs3org/reva/pull/1209) and is currently driven by CERN. 
+
+## Architecture
+
+In the original approach the driver was based on the localfs driver, relying on a locally mounted cephfs. It would interface with it using the POSIX apis. This has been changed to directly call the Ceph API using https://github.com/ceph/go-ceph. It allows using the ceph admin APIs to create subvolumes for user homes and maintain a file id to path mapping using symlinks.
+
+## Implemented Aspects
+The recursive change time built ino cephfs is used to implement the etag propagation expected by the ownCloud clients. This allows oCIS to pick up changes that have been made by external tools, bypassing any oCIS APIs. 
+
+Like other filesystems cephfs uses inodes and like most other filesystems inodes are reused. To get stable file identifiers the current cephfs driver assigns every node a file id and maintains a custom fileid to path mapping in a system directory:
```
Comment (dragotin) [inline docs/ocis/storage-backends/cephfs.md:25]: AGain, I think that is a bit brief. 

Questions people might come up with and which can be answered here:
- Why is it a problem if inodes are reused?
- What is an inode and what is ownCloud using it for?
- Where does the Ceph driver has a "file Id" from and why is that better?

File: docs/ocis/storage-backends/cephfs.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+Shares [are be mapped to ACLs](https://github.com/cs3org/reva/pull/1209/files#diff-5e532e61f99bffb5754263bc6ce75f84a30c6f507a58ba506b0b487a50eda1d9R168-R224) supported by cephfs. The share manager is used to persist the intent of a share and can be used to periodically verify or reset the ACLs on cephfs.
+
+## Future work
+- The spaces concept matches cephfs subvolumes. We can implement the CreateStorageSpace call with that, keep track of the list of storage spaces using symlinks, like for the id based lookup.
+- The share manager needs a persistence layer.
+- Currently we persist using a single json file.
+- As it basically provides two lists, *shared with me* and *shared with others*, we could persist them directly on cephfs!
+  - If needed for redundancy, the share manager can be run multiple times, backed by the same cephfs
+  - To save disk io the data can be cached in memory, and invalidated using stat requests.
```
Comment (dragotin) [inline docs/ocis/storage-backends/cephfs.md:53]: This is a implementation detail that I would not mention here.

File: docs/ocis/storage-backends/cephfs.md [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+Trash is not implemented, as cephfs has no native recycle bin and instead relies on the snapshot functionality that can be triggered by end users. It should be possible to automatically create a snapshot before deleting a file. This needs to be explored.
+
+Shares [are be mapped to ACLs](https://github.com/cs3org/reva/pull/1209/files#diff-5e532e61f99bffb5754263bc6ce75f84a30c6f507a58ba506b0b487a50eda1d9R168-R224) supported by cephfs. The share manager is used to persist the intent of a share and can be used to periodically verify or reset the ACLs on cephfs.
+
+## Future work
+- The spaces concept matches cephfs subvolumes. We can implement the CreateStorageSpace call with that, keep track of the list of storage spaces using symlinks, like for the id based lookup.
+- The share manager needs a persistence layer.
+- Currently we persist using a single json file.
```
Comment (dragotin) [inline docs/ocis/storage-backends/cephfs.md:50]: Maybe rather: _"The share manager uses a file system based persistance layer that is scaled accordingly. For single user scope as user shares, two files `shared_with_me.json` and `shared_with_others.json` can be considered sufficient, as the amount of shares per user does not grow infinite."_

File: docs/ocis/storage-backends/cephfs.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+│   │   ├── 50BC39D364A4703A20C58ED50E4EADC3_570078 -> /tmp/cephfs/reva/einstein
+│   │   ├── 571EFB3F0ACAE6762716889478E40156_570081 -> /tmp/cephfs/reva/einstein/Pictures
+│   │   └── C7A1397524D0419B38D04D539EA531F8_588108 -> /tmp/cephfs/reva/einstein/welcome.txt
+│   └── .uploads
+└── .reva_share_manager
+    ├── einstein
+    │   └── sharedWithOthers.json
+    └── marie
+        └── sharedWithMe.json
+```
```
Comment (dragotin) [inline docs/ocis/storage-backends/cephfs.md:76]: Same here - do not use `/tmp` at top level.

**Review by dragotin — APPROVED** [review]
A few comments and wishes for changes, but given the time pressure I am good.

**Pull request merged**